xarray/backends/zarr.py

```diff
@@ -178,19 +178,40 @@ def _determine_zarr_chunks(enc_chunks, var_chunks, ndim, name, safe_chunks):
     raise AssertionError("We should never get here. Function logic must be wrong.")
 
 
+def _get_nczarr_dims(zarr_obj):
+    # NCZarr defines dimensions through metadata in .zarray
+    zarray_path = os.path.join(zarr_obj.path, ".zarray")
+    zarray = zarr.util.json_loads(zarr_obj.store[zarray_path])
+    # NCZarr uses Fully Qualified Names
+    dimensions = [os.path.basename(dim) for dim in zarray["_NCZARR_ARRAY"]["dimrefs"]]
+    return dimensions
+
+
+def _hide_nczarr_attrs(attrs):
+    return HiddenKeyDict(attrs, [attr for attr in attrs if attr.startswith("_NC")])
+
+
 def _get_zarr_dims_and_attrs(zarr_obj, dimension_key):
     # Zarr arrays do not have dimensions. To get around this problem, we add
     # an attribute that specifies the dimension. We have to hide this attribute
     # when we send the attributes to the user.
     # zarr_obj can be either a zarr group or zarr array
+    attributes = _hide_nczarr_attrs(zarr_obj.attrs)
     try:
+        # Xarray-Zarr
         dimensions = zarr_obj.attrs[dimension_key]
     except KeyError:
-        raise KeyError(
-            f"Zarr object is missing the attribute `{dimension_key}`, which is "
-            "required for xarray to determine variable dimensions."
-        )
-    attributes = HiddenKeyDict(zarr_obj.attrs, [dimension_key])
+        try:
+            # NCZarr
+            dimensions = _get_nczarr_dims(zarr_obj)
+            attributes = dict(attributes)
+            attributes[dimension_key] = dimensions
```

Comment:
HiddenKeyDict is a thin proxy, so this is modifying the original attributes dict on the Zarr object. That seems a little unsafe -- could we avoid modifying attributes? I'm pretty sure HiddenKeyDict will work OK with some keys not being found.

Reply:
I'm not adding the `_ARRAY_DIMENSIONS` attribute anymore, and I'm using `_get_zarr_dims_and_attrs` in `self.get_dimensions` as well. This way, `_get_zarr_dims_and_attrs` is the only place where we look for `_ARRAY_DIMENSIONS` and NCZarr attributes